pkg/csi/service/vanilla/controller.go

```diff
@@ -2236,7 +2236,7 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 			nodevm, err = c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 		}
 		if err != nil {
-			if err == cnsvsphere.ErrVMNotFound {
+			if err == node.ErrNodeNotFound {
```

Comment:
Here we do not need to check `if err == node.ErrNodeNotFound {`

When `UseCSINodeId` feature is enabled, we do lookup the Node VM using `c.nodeMgr.GetNodeByName(ctx, req.NodeId)` first. If this function returns `node.ErrNodeNotFound `, we do look up again using `c.nodeMgr.GetNodeByUuid(ctx, req.NodeId)`, which will return `cnsvsphere.ErrVMNotFound` if Node is not present in the vCenter inventory.

We will be removing the feature gate check for `commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.UseCSINodeId)` and will be enabling `UseCSINodeId` feature by default, so the else part does not need to be corrected on the master branch. 

Also, we do not want to mark detach as successful without checking vCenter inventory to ensure VM is really not there in the vCenter inventory and it is safe to mark the volume as detached.

Reply:
Wow. I've missed that completely. Thanks for detailed response! It looks like I've missed something that makes detaches unreliable.

Reply:
Thanks again! Will wait for a proper fix!